eras/mary/impl/testlib/Test/Cardano/Ledger/Mary/Arbitrary.hs

```diff
@@ -89,9 +98,88 @@ instance Crypto c => Arbitrary (PolicyID c) where
+--      not exceed 65535.
+--   5. With `n` as the total number of assets, `p` the number of policy ids, the inequality to be satisfied is thus:
+--           8n -- Word64 asset quantities
+--        +  2n -- Word16 policy id offsets
+--        +  2n -- Word16 asset name offsets
+--        + 28p -- 28-byte policy ids
+--        + 32n -- 32-byte asset names (a maximum of 32 bytes)
+--        should be <= 65535, assuming the numer of policies to be maximal (i.e. equal to number of assets)
+--        65535 / 72 ~ 910.2 is the maximum number of triples to be safely generated
+--
+-- NOTE: There are some conditions due to which exceeding this number may not
+-- result in a guaranteed failure to compact without overflow, because, during compacting
+--   1. The asset names and policy ids are deduplicated
+--   2. Not all generated asset names are 32-bytes long
+-- But, exceeding this number does make the probability of causing overflow > 0.
+genMultiAsset :: forall c. Crypto c => Gen Integer -> Gen (MultiAsset c)
+genMultiAsset genAmount = do
+  MultiAsset ma <-
+    oneof
+      [ MultiAsset <$> genNonEmptyMap arbitrary (genNonEmptyMap arbitrary genAmount)
+      , multiAssetFromListBounded <$> listOf1 (genMultiAssetTriple $ fromIntegral <$> genAmount)
+      ]
+  if sum (length <$> ma) > 910
+    then scale (`div` 2) $ genMultiAsset genAmount
+    else pure $ MultiAsset ma
+
+-- | For tests, because `insertMultiAsset` called through `genMultiAsset` filters out zero values
+genMultiAssetZero :: Crypto c => Gen (MultiAsset c)
+genMultiAssetZero = MultiAsset <$> genNonEmptyMap arbitrary (genNonEmptyMap arbitrary (pure 0))
+
+-- For negative tests, we need a definite generator that causes overflow
+genMultiAssetToFail :: forall c. Crypto c => Gen (MultiAsset c)
+genMultiAssetToFail = do
+  (numPolicyId, numAssetName) <-
+    oneof
+      [ do
+          Positive numAsssetNames <- arbitrary
+          let minNumPolicyId = (65535 - 44 * numAsssetNames) `div` 28
+          numPolicyIds <- (minNumPolicyId +) . getPositive <$> arbitrary
+          pure (numPolicyIds, numAsssetNames)
+      , do
+          Positive numPolicyIds <- arbitrary
+          let minNumAssetNames = (65535 - 28 * numPolicyIds) `div` 44
+          numAssetNames <- (minNumAssetNames +) . getPositive <$> arbitrary
+          pure (numPolicyIds, numAssetNames)
+      ]
+  -- Here we generate separately a list of asset names and a list of policy ids and
+  -- randomly shuffle them into a MultiAsset
+
+  if numPolicyId > numAssetName -- Since we don't want to generate policies with empty assets
+    then genMultiAssetToFail
+    else do
+      ps <- resize numPolicyId $ vectorOf numPolicyId (arbitrary :: Gen (PolicyID c))
+      as <- resize numAssetName $ vectorOf numAssetName (arbitrary :: Gen (AssetName, Int))
+      let
+        initialTriples = zipWith (\p (a, v) -> (p, a, v)) ps as -- All policies should have at least one asset
```

Comment:
`zipWith` will drop generated elements. In other words `length initialTriples == min (length ps) (length as)

So this whole thing is wrong

Reply:
> Noting for posterity: This part was re-reviewed and it was decided that with minor adjustments, overall it works well.

Reply:
Yes. It was my mistake. This whole thing is totally right :slightly_smiling_face: